File: linux/hutuu/src/go/pipe.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"syscall"
)

func main() {
	if len(os.Args) != 3 {
		fmt.Fprintf(os.Stderr, "Usage: %s <command1> <command2>\n", os.Args[0])
		os.Exit(1)
	}

	rfd, wfd, err := os.Pipe()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Pipe creation failed: %v\n", err)
		os.Exit(1)
	}

	path1, err := exec.LookPath(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to find %s: %v\n", os.Args[1], err)
		os.Exit(1)
	}

	pid, err := syscall.ForkExec(path1, []string{os.Args[1]}, &syscall.ProcAttr{
		Dir:   ".",
		Env:   os.Environ(),
		Files: []uintptr{os.Stdin.Fd(), uintptr(wfd.Fd()), os.Stderr.Fd()},
		Sys:   nil,
	})
	if err != nil {
		fmt.Fprintf(os.Stderr, "ForkExec failed: %v\n", err)
		os.Exit(1)
	}

	wfd.Close()

	path2, err := exec.LookPath(os.Args[2])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to find %s: %v\n", os.Args[2], err)
		os.Exit(1)
	}

	syscall.Dup3(int(rfd.Fd()), int(os.Stdin.Fd()), 0)
	rfd.Close()

	err = syscall.Exec(path2, []string{os.Args[2]}, os.Environ())
	if err != nil {
		fmt.Fprintf(os.Stderr, "Exec failed: %v\n", err)
		os.Exit(1)
	}

	syscall.Wait4(pid, nil, 0, nil)
}

File: linux/hutuu/src/pipe.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>

int main(int argc, char **argv)
{
    if (argc != 3)
    {
        fprintf(stderr, "Usage: %s <command1> <command2>\n", argv[0]);
        exit(1);
    }

    int fds[2];
    // 両端とも自プロセスに繋がったストリームを生成する。
    // fds[0] は読み込み専用 fds[1] は書き込み専用。
    pipe(fds);

    // 子プロセスと親プロセスの両方で呼び出しが戻る！！
    int pid = fork();
    if (pid < 0)
    {
        fprintf(stderr, "fork(2) failed\n");
        exit(1);
    }

    if (pid == 0)
    {
        // 子プロセス。

        // 書き込みのストリーム用の fd を閉じる。
        close(fds[1]);
        // 読み込みを標準入力のストリームに複製する。
        dup2(fds[0], STDIN_FILENO);
        close(fds[0]);

        execlp(argv[2], argv[2], NULL);

        // execlpが失敗した場合。
        perror("execlp child");
        exit(1);
    }
    else
    {
        // 親プロセス。

        // 読み込みのストリーム用の fd を閉じる。
        close(fds[0]);
        // 書き込みを標準出力のストリームに複製する。
        dup2(fds[1], STDOUT_FILENO);
        close(fds[1]);

        execlp(argv[1], argv[1], NULL);

        // execlpが失敗した場合。
        perror("execlp parent");
        exit(1);
    }

    wait(NULL);

    exit(0);
}
